consumer.go

```diff
@@ -89,298 +145,453 @@ func NewConsumer(client *Client, topic string, partition int32, group string, co
+	// is in the middle of a network request) and batches it up. The main worker goroutine picks
+	// up a batch of new work between every network request by reading from `newWork`, so we give
+	// it nil if no new work is available. We also write to `wait` only when new work is available,
+	// so the main goroutine can block waiting for work if it has none.
+	for {
+		if len(buffer) > 0 {
+			select {
+			case event, ok := <-w.input:
+				if !ok {
+					goto done
 				}
+				buffer = append(buffer, event)
+			case w.newWork <- buffer:
+				buffer = nil
+			case w.wait <- none{}:
 			}
+		} else {
 			select {
-			case <-c.stopper:
-				close(c.events)
-				close(c.done)
-				return
-			default:
-				continue
+			case event, ok := <-w.input:
+				if !ok {
+					goto done
+				}
+				buffer = append(buffer, event)
+			case w.newWork <- nil:
 			}
-		} else {
-			fetchSize = c.config.DefaultFetchSize
 		}
+	}
 
-		atLeastOne := false
-		for _, msgBlock := range block.MsgSet.Messages {
-			prelude := true
+done:
+	close(w.wait)
+	if len(buffer) > 0 {
+		w.newWork <- buffer
+	}
+	close(w.newWork)
+}
 
-			for _, msg := range msgBlock.Messages() {
-				if prelude && msg.Offset < c.offset {
-					continue
-				}
-				prelude = false
-
-				event := &ConsumerEvent{Topic: c.topic, Partition: c.partition}
-				if msg.Offset == c.offset {
-					atLeastOne = true
-					event.Key = msg.Msg.Key
-					event.Value = msg.Msg.Value
-					event.Offset = msg.Offset
-					c.offset++
-				} else {
-					event.Err = IncompleteResponse
-				}
+func (w *consumerWorker) doWork() {
+	<-w.wait // wait for our first piece of work
 
-				select {
-				case <-c.stopper:
-					close(c.events)
-					close(c.done)
-					return
-				case c.events <- event:
-				}
-			}
+	// the bridge ensures we will get nil right away if no new work is available
+	for newWork := range w.newWork {
+		w.updateWorkCache(newWork)
 
+		if len(w.work) == 0 {
+			// We're about to be shut down or we're about to receive more work.
+			// Either way, the signal just hasn't propagated to our goroutine yet.
+			<-w.wait
+			continue
 		}
 
-		if !atLeastOne {
-			select {
-			case <-c.stopper:
-				close(c.events)
-				close(c.done)
-				return
-			case c.events <- &ConsumerEvent{Topic: c.topic, Partition: c.partition, Err: IncompleteResponse}:
+		response, err := w.fetchNewMessages()
+
+		if err != nil {
+			Logger.Printf("Unexpected error processing FetchRequest; disconnecting broker %s: %s\n", w.broker.addr, err)
+			w.abort(err)
+			return
+		}
+
+		for child, _ := range w.work {
+			block := response.GetBlock(child.topic, child.partition)
+			if block == nil {
+				// TODO should we be doing anything else here? The fact that we didn't get a block at all
+				// (not even one with an error) suggests that the broker is misbehaving, or perhaps something in the
+				// request/response pipeline is ending up malformed, so we could be better off aborting entirely...
```

Comment:
I was just thinking about the comment there, that maybe we should abort, but what if there are other valid blocks in the response. But it seems that right now we won't detect the case when there isn't any valid block in the response. Would we want to abort in that case?